File: app/src/main/java/com/nfcalarmclock/alarm/db/NacAlarmDao.java
package com.nfcalarmclock.alarm.db;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;
import java.util.List;

/**
 * Data access object for alarms.
 */
@Dao
public interface NacAlarmDao
{

	/**
	 * Delete an alarm.
	 *
	 * @param  alarm  Alarm to delete.
	 *
	 * @return The number of rows deleted.
	 */
	@Delete
	int delete(NacAlarm alarm);

	/**
	 * Delete all alarms.
	 *
	 * @return The number of rows deleted.
	 */
	@Query("DELETE FROM alarm")
	int deleteAll();

	/**
	 * Find an alarm.
	 *
	 * @param  id  The ID of the alarm to find.
	 *
	 * @return The alarm with the ID.
	 */
	@Query("SELECT * FROM alarm WHERE id=:id")
	NacAlarm findAlarm(long id);

	/**
	 * Get an active alarm.
	 *
	 * @return An active alarm.
	 */
	@Query("SELECT * FROM alarm WHERE is_active=1 LIMIT 1")
	LiveData<NacAlarm> getActiveAlarm();

	/**
	 * Get a list of all active alarms.
	 *
	 * @return List of all active alarms.
	 */
	@Query("SELECT * FROM alarm WHERE is_active=1")
	LiveData<List<NacAlarm>> getActiveAlarms();

	/**
	 * Get a list of all active alarms.
	 * <p>
	 * This will wait until all alarms are selected.
	 *
	 * @return List of all active alarms.
	 */
	@Query("SELECT * FROM alarm WHERE is_active=1")
	List<NacAlarm> getActiveAlarmsNow();

	/**
	 * Get all alarms.
	 *
	 * @return All alarms.
	 */
	@Query("SELECT * FROM alarm")
	LiveData<List<NacAlarm>> getAllAlarms();

	/**
	 * Get all alarms.
	 * <p>
	 * This will wait until all alarms are selected.
	 *
	 * @return All alarms.
	 */
	@Query("SELECT * FROM alarm")
	List<NacAlarm> getAllAlarmsNow();

	/**
	 * Insert an alarm.
	 *
	 * @param  alarm  The alarm to insert.
	 *
	 * @return The row ID of the alarm that was inserted.
	 */
	@Insert()
	long insert(NacAlarm alarm);

	/**
	 * Update an existing alarm
	 *
	 * @param  alarm  The alarm to update.
	 *
	 * @return The number of alarms updated.
	 */
	@Update
	int update(NacAlarm alarm);

}

File: app/src/main/kotlin/com/nfcalarmclock/alarm/db/NacAlarmDao.kt
package com.nfcalarmclock.alarm.db

import androidx.lifecycle.LiveData
import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.Query
import androidx.room.RawQuery
import androidx.room.Update
import androidx.sqlite.db.SupportSQLiteQuery

/**
 * Data access object for alarms.
 */
@Dao
interface NacAlarmDao
{

	/**
	 * Get all alarms.
	 *
	 * @return All alarms.
	 */
	@get:Query("SELECT * FROM alarm")
	val allAlarmsLive: LiveData<List<NacAlarm>>

	/**
	 * Checkpoint the database, which will block until there is no database writer and
	 * all readers are reading from the most recent database snapshot.
	 *
	 * It then checkpoints all frames in the log file and syncs the database file.
	 *
	 * This makes it so that the database does not need to be closed before exporting it.
	 */
	@RawQuery
	suspend fun checkpoint(supportSQLiteQuery: SupportSQLiteQuery): Int

	/**
	 * Count the number of alarms.
	 *
	 * @return The number of alarms in the table.
	 */
	@Query("SELECT COUNT(id) FROM alarm")
	suspend fun count(): Int

	/**
	 * Delete an alarm.
	 *
	 * @param  alarm  Alarm to delete.
	 *
	 * @return The number of rows deleted.
	 */
	@Delete
	suspend fun delete(alarm: NacAlarm): Int

	/**
	 * Find an alarm.
	 *
	 * @param  id  The ID of the alarm to find.
	 *
	 * @return The alarm with the ID.
	 */
	@Query("SELECT * FROM alarm WHERE id=:id")
	suspend fun findAlarm(id: Long): NacAlarm?

	/**
	 * Get an active alarm, the alarm that has been active the longest.
	 *
	 * @return An active alarm, the alarm that has been active the longest.
	 */
	@Query("SELECT * FROM alarm WHERE is_active=1 ORDER BY time_active DESC LIMIT 1")
	suspend fun getActiveAlarm(): NacAlarm?

	/**
	 * Get a list of all active alarms.
	 * <p>
	 * This will wait until all alarms are selected.
	 *
	 * @return List of all active alarms.
	 */
	@Query("SELECT * FROM alarm WHERE is_active=1")
	suspend fun getActiveAlarms(): List<NacAlarm>

	/**
	 * Get all alarms.
	 * <p>
	 * This will wait until all alarms are selected.
	 *
	 * @return All alarms.
	 */
	@Query("SELECT * FROM alarm")
	suspend fun getAllAlarms(): List<NacAlarm>

	/**
	 * Insert an alarm.
	 *
	 * @param  alarm  The alarm to insert.
	 *
	 * @return The row ID of the alarm that was inserted.
	 */
	@Insert
	suspend fun insert(alarm: NacAlarm): Long

	/**
	 * Update an existing alarm.
	 *
	 * @param  alarm  The alarm to update.
	 *
	 * @return The number of alarms updated.
	 */
	@Update
	suspend fun update(alarm: NacAlarm): Int

}